### PR TITLE
Add type hints to returnn/compile.py

### PR DESCRIPTION
**Pull request by Icemole**
The requirements were hardcoded to None, so it was a bit hard to achieve something from this job other than for it to be a mini task.

### CODE REVIEW

**Comment by Atticus1806** [comment]
Good Idea but I think you need to add a hash exclude similar to the other two for the default in order for it not to break hashes.
Also thanks for the typing refactoring :)

**Review by kuacakuaca — APPROVED** [review]

**Review by kuacakuaca — CHANGES_REQUESTED** [review]
request changes because of hash break

File: returnn/compile.py [hunk 1]
```diff
@@ -2,62 +2,60 @@
...
-        verbosity=4,
-        device=None,
+        returnn_config: Union[ReturnnConfig, tk.Path, str],
+        train: int = 0,
+        eval: int = 0,
+        search: int = 0,
+        epoch: Optional[Union[int, tk.Variable]] = None,
+        verbosity: int = 4,
+        device: Optional[str] = None,
         summaries_tensor_name=None,
```
Comment (michelwi) [inline returnn/compile.py:35]:
```suggestion
        summaries_tensor_name: str =None,
```
I think this should be str?

File: returnn/compile.py [hunk 2]
```diff
@@ -2,62 +2,60 @@
...
-        output_format="meta",
-        returnn_python_exe=None,
-        returnn_root=None,
-        rec_step_by_step=None,
-        rec_json_info=False,
+        output_format: str = "meta",
+        returnn_python_exe: Optional[tk.Path] = None,
+        returnn_root: Optional[tk.Path] = None,
+        rec_step_by_step: Optional[tk.Path] = None,
```
Comment (michelwi) [inline returnn/compile.py:39]:
```suggestion
        rec_step_by_step: Optional[str] = None,
```
I think the old typing was correct

**Comment by Icemole** [comment]
@Atticus1806 @michelwi @kuacakuaca I've already fixed this, thanks for warning me.

I've also done a bit more type refactoring since you liked it :)

File: returnn/compile.py (hunk 1 above)
Comment (Icemole) [reply returnn/compile.py:35]: I guess so, but I didn't see any example of this.

I'll just set it as string :woman_shrugging:

File: returnn/compile.py (hunk 2 above)
Comment (Icemole) [reply returnn/compile.py:39]: Definitely. Copy paste issue :sweat:

**Review by Atticus1806 — APPROVED** [review]

**Comment by curufinwe** [comment]
Iirc having rqmt as a parameter is discouraged. You can just update the resulting job object to set new requirements. This is done in many other jobs like that.

**Comment by Icemole** [comment]
@curufinwe suggested me to just set the created job's requirements manually through `job.rqmt = ...`. Would that be something you'd agree on? If so, maybe this PR could be just about adding some typing hints and cleaning up the code a bit ^^

**Comment by Icemole** [comment]
I've just reverted the code and changed the title accordingly :slightly_smiling_face:

**Review by michelwi — APPROVED** [review]

**Pull request merged**